File: MinecraftPvpPlugin/src/main/java/me/minecraft/minecraftpvpplugin/gadgets/Rebound.java
package me.minecraft.minecraftpvpplugin.gadgets;

import me.minecraft.minecraftpvpplugin.Gadget;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.player.PlayerInteractAtEntityEvent;
import org.bukkit.event.player.PlayerInteractEvent;

public class Rebound extends Gadget implements Listener {
    public Rebound() {
        super(Material.WOOD_DOOR, "反射之盾", 10);
    }
    public void onActivate(PlayerInteractEvent event) {
        Player player = event.getPlayer();
        player.sendMessage("start");
    }
    @EventHandler
    public void EntityDamageByEntityEvent(EntityDamageByEntityEvent event){
        Player player = (Player) event.getDamager();
        if(PlayersUsingGadget.contains((Player) event.getEntity())){
            int health = (int) player.getHealth();
            int hurt = (int) event.getDamage();
            player.setHealth(health - (double) hurt /2);
        }
    }

    public void onDeactivate(PlayerInteractEvent event) {
        Player player = event.getPlayer();
        player.sendMessage("stop");
    }
}

File: MinecraftPvpPlugin/src/main/java/me/minecraft/minecraftpvpplugin/gadgets/Rebound.kt
package me.minecraft.minecraftpvpplugin.gadgets

import me.minecraft.minecraftpvpplugin.CustomEffect
import me.minecraft.minecraftpvpplugin.Gadget
import me.minecraft.minecraftpvpplugin.LogWriter
import org.bukkit.ChatColor
import org.bukkit.Effect
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent

object Rebound : Gadget(Material.WOOD_DOOR, "反射之盾", duration = 10.0, lore = listOf(
    "${ChatColor.YELLOW}反彈所受傷害的50%",
    "${ChatColor.GRAY}持續時間：10秒"
)), Listener {
    @EventHandler
    fun entityDamageByEntityEvent(event: EntityDamageByEntityEvent) {
        if (event.damager is Player) {
            val damager = event.damager as Player
            val damaged = event.entity as Player

            if (isActivating(damaged)) {
                LogWriter.log("${damaged.name} use 反射之盾 against ${damager.name}")

                CustomEffect.playParticleInSphere(damaged, Effect.FIREWORKS_SPARK, 100, 0.8f)
                val health = damager.health.toInt()
                val hurt = event.finalDamage.toInt()
                damager.health = health - hurt.toDouble() / 2
            }
        }
    }
}
